### PR TITLE
Moves sensor data transformation into SensorBridge.

### PR DESCRIPTION
**Pull request by damonkohler**
* Pulls out TfBridge.
* SensorBridge has its own options.
* SensorBridge is now responsible for transforming the sensor data into the tracking frame.
* Constant odometry covariance is no longer optional.

### CODE REVIEW

File: cartographer_ros/src/node_options.h [hunk 1]
```diff
@@ -23,25 +23,22 @@
...
 #include "cartographer/mapping/map_builder.h"
 
+#include "sensor_bridge.h"
+
 namespace cartographer_ros {
 
 // Top-level options of Cartographer's ROS integration.
 struct NodeOptions {
   ::cartographer::mapping::proto::MapBuilderOptions map_builder_options;
+  SensorBridgeOptions sensor_bridge_options;
```
Comment (wohe) [inline cartographer_ros/src/node_options.h]: Update the documentation by prefixing `sensor_bridge.` in the appropriate places?

File: cartographer_ros/src/sensor_bridge.h [hunk 2]
```diff
@@ -28,32 +30,53 @@
...
+                                const sensor_msgs::PointCloud2::ConstPtr& msg);
 
  private:
+  void HandleLaserScanProto(
+      const string& topic, const ::cartographer::common::Time time,
+      const string& frame_id,
+      const ::cartographer::sensor::proto::LaserScan& laser_scan);
+
+  const SensorBridgeOptions options_;
+  const TfBridge* tf_bridge_;
```
Comment (wohe) [inline cartographer_ros/src/sensor_bridge.h]: `* const` here and below?

File: cartographer_ros/src/sensor_bridge.cc [hunk 3]
```diff
@@ -16,75 +16,148 @@
...
+      lua_parameter_dictionary->GetDouble(
+          "constant_odometry_translational_variance");
+  options.constant_odometry_rotational_variance =
+      lua_parameter_dictionary->GetDouble(
+          "constant_odometry_rotational_variance");
+  return options;
+}
+
 SensorBridge::SensorBridge(
+    const SensorBridgeOptions& options, const TfBridge* tf_bridge,
```
Comment (wohe) [inline cartographer_ros/src/sensor_bridge.cc]: `* const`?

File: cartographer_ros/src/sensor_bridge.cc [hunk 4]
```diff
@@ -16,75 +16,148 @@
...
-                           ToPoseCovariance(msg->pose.covariance)});
-  sensor_collator_->AddSensorData(
-      trajectory_id_,
-      ::cartographer::common::ToUniversal(FromRos(msg->header.stamp)), topic,
-      std::move(sensor_data));
+void SensorBridge::HandleOdometryMessage(
+    const string& topic, const nav_msgs::Odometry::ConstPtr& msg) {
+  const carto::common::Time time = FromRos(msg->header.stamp);
+  carto::kalman_filter::PoseCovariance applied_covariance =
+      ToPoseCovariance(msg->pose.covariance);
```
Comment (wohe) [inline cartographer_ros/src/sensor_bridge.cc]: Move to the else branch. Also: Doesn't this need to be transformed?

File: cartographer_ros/src/sensor_bridge.cc [hunk 5]
```diff
@@ -16,75 +16,148 @@
...
 
-void SensorBridge::AddPointCloud2Message(
+void SensorBridge::HandlePointCloud2Message(
     const string& topic, const sensor_msgs::PointCloud2::ConstPtr& msg) {
-  pcl::PointCloud<pcl::PointXYZ> pcl_points;
-  pcl::fromROSMsg(*msg, pcl_points);
+  pcl::PointCloud<pcl::PointXYZ> pcl_point_cloud;
+  pcl::fromROSMsg(*msg, pcl_point_cloud);
+  const auto sensor_to_tracking = tf_bridge_->LookupToTracking(
+      FromRos(msg->header.stamp), msg->header.frame_id);
```
Comment (wohe) [inline cartographer_ros/src/sensor_bridge.cc]: Consistency: `const carto::common::Time time = FromRos(msg->header.stamp);` first?

File: cartographer_ros/src/tf_bridge.cc [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "cartographer/common/make_unique.h"
+
+#include "msg_conversion.h"
+#include "tf_bridge.h"
+
+namespace cartographer_ros {
+
+TfBridge::TfBridge(const string& tracking_frame,
+                   double lookup_transform_timeout_sec,
```
Comment (damonkohler) [inline cartographer_ros/src/tf_bridge.cc]: Done.

File: cartographer_ros/src/tf_bridge.cc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            ->lookupTransform(tracking_frame_, frame_id, ::ros::Time(0.),
+                              timeout)
+            .header.stamp;
+    const ::ros::Time requested_time = ToRos(time);
+    if (latest_tf_time >= requested_time) {
+      // We already have newer data, so we do not wait. Otherwise, we would wait
+      // for the full 'timeout' even if we ask for data that is too old.
+      timeout = ::ros::Duration(0.);
+    }
+    frame_id_to_tracking =
```
Comment (damonkohler) [inline cartographer_ros/src/tf_bridge.cc]: Done.

File: cartographer_ros/src/node_options.h (hunk 1 above)
Comment (damonkohler) [inline cartographer_ros/src/node_options.h]: Done.

File: cartographer_ros/src/sensor_bridge.h (hunk 2 above)
Comment (damonkohler) [inline cartographer_ros/src/sensor_bridge.h]: Done.

File: cartographer_ros/src/sensor_bridge.cc (hunk 3 above)
Comment (damonkohler) [inline cartographer_ros/src/sensor_bridge.cc]: Done

File: cartographer_ros/src/sensor_bridge.cc (hunk 5 above)
Comment (damonkohler) [inline cartographer_ros/src/sensor_bridge.cc]: Done.

File: cartographer_ros/src/cartographer_node_main.cc [hunk 8]
```diff
@@ -615,23 +503,28 @@ void Node::HandleSensorData(const int64 timestamp,
     last_sensor_data_rates_logging_time_ = std::chrono::steady_clock::now();
   }
 
+  const auto time = carto::common::FromUniversal(timestamp);
+  auto trajectory_builder =
```
Comment (wohe) [inline cartographer_ros/src/cartographer_node_main.cc]: `auto *`?

File: cartographer_ros/src/sensor_bridge.cc [hunk 9]
```diff
@@ -16,75 +16,143 @@
...
+    CHECK(sensor_to_tracking->translation().norm() < 1e-5)
+        << "The IMU frame must be colocated with the tracking frame. "
+           "Transforming linear acceleration into the tracking frame will "
+           "otherwise be imprecise.";
+    sensor_collator_->AddSensorData(
+        trajectory_id_, carto::common::ToUniversal(time), topic,
+        carto::common::make_unique<SensorData>(
+            msg->header.frame_id,
+            SensorData::Imu{
+                sensor_to_tracking->rotation() * ToEigen(msg->angular_velocity),
```
Comment (wohe) [inline cartographer_ros/src/sensor_bridge.cc]: Swap angular velocity/linear accelerations for consistency: The variance options, the covariance matrix and the cartographer library all have the opposite ordering.

File: docs/source/configuration.rst [hunk 10]
```diff
@@ -55,15 +55,11 @@ use_odometry_data
...
   must be provided in this case, and the information will be included in SLAM.
 
-use_constant_odometry_variance
-  If enabled, the configured variances will be used instead of the ones provided
-  in the Odometry messages.
-
-constant_odometry_translational_variance
+sensor_bridge.constant_odometry_translational_variance
   The variance to use for the translational component of odometry to use if
   *use_constant_odometry_variance* is enabled.
```
Comment (wohe) [inline docs/source/configuration.rst]: ..."if odometry is used"? Same in other places since the option is now gone.

Comment (damonkohler) [inline docs/source/configuration.rst]: Done.

File: cartographer_ros/src/sensor_bridge.cc (hunk 9 above)
Comment (damonkohler) [inline cartographer_ros/src/sensor_bridge.cc]: Done.

File: cartographer_ros/src/cartographer_node_main.cc (hunk 8 above)
Comment (damonkohler) [inline cartographer_ros/src/cartographer_node_main.cc]: Done.

**Pull request merged**